test/store_unittest.cc

```diff
@@ -0,0 +1,229 @@
+  const auto metadata_map_after = GetMetadataMap();
+  EXPECT_EQ(1, metadata_map_after.size());
+  EXPECT_EQ("default-version2", metadata_map_after.at(mr).version);
+}
+
+TEST_F(MetadataStoreTest, PurgeDeletedEntriesDeletesCorrectMetadata) {
+  MonitoredResource mr1("type1", {});
+  MonitoredResource mr2("type2", {});
+  MetadataStore::Metadata m1(
+      "default-version1",
+      false,
+      std::chrono::system_clock::now(),
+      std::chrono::system_clock::now(),
+      json::object({{"f", json::string("hello")}}));
+  MetadataStore::Metadata m2(
+      "default-version2",
+      true,
+      std::chrono::system_clock::now(),
+      std::chrono::system_clock::now(),
+      json::object({{"f", json::string("hello")}}));
+  store.UpdateMetadata(mr1, std::move(m1));
+  store.UpdateMetadata(mr2, std::move(m2));
+  const auto metadata_map_before = GetMetadataMap();
+  EXPECT_EQ(2, metadata_map_before.size());
+  EXPECT_EQ("default-version1", metadata_map_before.at(mr1).version);
+  EXPECT_EQ("default-version2", metadata_map_before.at(mr2).version);
+  PurgeDeletedEntries();
+  const auto metadata_map_after = GetMetadataMap();
+  EXPECT_EQ(1, metadata_map_after.size());
+  EXPECT_EQ("default-version1", metadata_map_after.at(mr1).version);
+  EXPECT_THROW(metadata_map_after.at(mr2), std::out_of_range);
+}
+
+TEST(MetadataTest, MetadataCorrectlyConstructed) {
+  MetadataStore::Metadata m(
+      "default-version",
+      false,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::object({{"f", json::string("hello")}}));
+  EXPECT_FALSE(m.ignore);
+  EXPECT_EQ("default-version", m.version);
+  EXPECT_FALSE(m.is_deleted);
+  EXPECT_EQ(time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+            m.created_at);
+  EXPECT_EQ(time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+            m.collected_at);
+  EXPECT_EQ("{\"f\":\"hello\"}", m.metadata->ToString());
+}
+
+TEST(MetadataTest, MetadataCorrectlyCloned) {
+  MetadataStore::Metadata m(
+      "default-version",
+      false,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::object({{"f", json::string("hello")}}));
+  MetadataStore::Metadata m_clone = m.Clone();
+  EXPECT_FALSE(m_clone.ignore);
+  EXPECT_EQ(m.version, m_clone.version);
+  EXPECT_FALSE(m_clone.is_deleted);
+  EXPECT_EQ(m.created_at, m_clone.created_at);
+  EXPECT_EQ(m.collected_at, m_clone.collected_at);
+  EXPECT_EQ(m.metadata->ToString(), m_clone.metadata->ToString());
+}
+
+TEST(MetadataTest, IgnoredMetadataCorrectlyCreated) {
+  MetadataStore::Metadata m = MetadataStore::Metadata::IGNORED();
+  EXPECT_TRUE(m.ignore);
+}
+
+TEST(MetadataTest, IgnoredMetadataCorrectlyCloned) {
+  MetadataStore::Metadata m = MetadataStore::Metadata::IGNORED();
+  MetadataStore::Metadata m_clone = m.Clone();
```

Comment:
It might seem like overkill, but it would be nice to add `EXPECT_TRUE(m.ignore);`, as I believe m.Clone() could potentially mutate the object, and thus copy from there and have both be false. If it's not possible that the value could be mutated, ignore :)

Reply:
So I was on the same train but @igorpeshansky 's point is that we should test that "cloning" an `IGNORED` metadata type works. As a result if `.ignore` before and after is the same the test worked, independent of the actual value. As is it is impossible to mutate the type of the `ignore` field because it can only be set via the default constructor.

Reply:
@dhrupadb, weren't you testing `m_clone.ignore`? Bryan suggests testing `m.ignore`. However, `Clone()` is declared `const`, so it shouldn't mutate any publicly visible state.

Reply:
Done. Added a check for `m`.

Reply:
No point in checking `m`, really, unless you don't trust `const`.
One thing that would help is changing the implementation of `Clone()` to use `IGNORED()` directly rather than the empty constructor/initializer list (effectively inlining the `IGNORED()` implementation). I've taken the liberty of pushing that change into your branch, hope that's ok.

Reply:
Yep LGTM. Thanks!